lib/internal/url.js

```diff
@@ -1365,13 +1365,16 @@ function getPathFromURLPosix(url) {
   return decodeURIComponent(pathname);
 }
 
-function getPathFromURL(path) {
+function getPathFromURL(path, assertFilename) {
   if (path == null || !path[searchParams] ||
       !path[searchParams][searchParams]) {
     return path;
   }
   if (path.protocol !== 'file:')
     throw new errors.TypeError('ERR_INVALID_URL_SCHEME', 'file');
+
+  if (assertFilename && path.href[path.href.length - 1] === '/')
+    throw new errors.Error('ERR_PATH_IS_DIRECTORY', 'path', path.pathname);
```

Comment:
The string being checked and reported here should be the result of  `isWindows ? getPathFromURLWin32(path) : getPathFromURLPosix(path)` for consistency, also because errors like `ERR_INVALID_FILE_URL_HOST`  should be thrown first.

Reply:
Or, maybe `getPathFromURL` does not even need to be changed since the path extracted from URLs will be eventually passed to `validatePath` anyway..from what I can tell if we update `WriteStream` and `ReadStream` to `validatePath(path, true)` then we don't need to a ssert the file name in `getPathFromURL`